spec/models/concern/mail_template_concern_spec.rb

```diff
@@ -104,6 +104,61 @@
         it { is_expected.to eq("--lien attestation--") }
       end
     end
+
+    describe '--lien document justificatif--' do
+      let(:procedure) { create(:procedure) }
+
+      subject { mail.body_for_dossier(dossier) }
+
+      before do
+        mail.body = "--lien document justificatif--"
+      end
+
+      describe "in closed mail" do
+        let(:mail) { create(:closed_mail, procedure: procedure) }
+        describe 'without justificatif' do
+          it { is_expected.to include("[l'instructeur n'a pas joint de document supplémentaire]") }
+        end
+
+        describe 'with justificatif' do
+          before do
+            dossier.justificatif_motivation.attach(justificatif)
+          end
+          it { expect(dossier.justificatif_motivation).to be_attached }
+          it { is_expected.to include("Télécharger le document justificatif") }
+        end
+      end
+
+      describe "in refused mail" do
+        let(:mail) { create(:refused_mail, procedure: procedure) }
+        describe 'without justificatif' do
+          it { is_expected.to include("[l'instructeur n'a pas joint de document supplémentaire]") }
+        end
+
+        describe 'with justificatif' do
+          before do
+            dossier.justificatif_motivation.attach(justificatif)
+          end
+          it { expect(dossier.justificatif_motivation).to be_attached }
+          it { is_expected.to include("Télécharger le document justificatif") }
+        end
+      end
+
+      describe "in without continuation mail" do
+        let(:mail) { create(:without_continuation_mail, procedure: procedure) }
+        describe 'without justificatif' do
+          it { is_expected.to include("[l'instructeur n'a pas joint de document supplémentaire]") }
+        end
+
+        describe 'with justificatif' do
+          before do
+            dossier.justificatif_motivation.attach(justificatif)
+          end
+          it { expect(dossier.justificatif_motivation).to be_attached }
+          it { is_expected.to include("Télécharger le document justificatif") }
+        end
+      end
+    end
```

Comment:
J'ai l'impression que ces trois tests se ressemblent beaucoup.

En ce qui me concerne, ça m'irait de tester seulement un seul des états (par exemple `closed_mail`) : j'ai l'impression que le rapport "couverture du test" sur "quantité de lignes nécessaire pour cette couverture", on serait meilleurs (on couvre un peu moins, mais en beauuucoup moins de lignes).

(Cela dit, si on tient vraiment à tester tous les états, on peut aussi faire des `shared_examples`.)

Reply:
les tests se ressemble totalement, mais une des autres caractéristiques du --lien attestation-- c'est qu'il n'est dispo que pour le passage en `accepté`. Ça ne me semble pas souhaitable je pense dans le cas présent, je voulais marquer le fait qu'on remplacait bien le tag dans les 3 états. Je peux aussi faire une boucle pour n'avoir qu'un seule corps de test, ou tout meilleur équivalent rspec de ce genre de choses.

Reply:
Apparement un bon compromis c'est de faire un [test paramétré](https://www.rubysteps.com/refactor-away-duplication-in-your-rspec-tests-using-the-parameterized-test-method) de cette manière:
```
describe '#double_it' do
  def self.test_double_it(initial, doubled)
    it "doubles #{initial} into #{doubled}" do
      expect(double_it(initial)).to eq(doubled)
    end
  end

  test_double_it 1, 2
  test_double_it 2, 4
  test_double_it 4, 8
end
```
Un seul corps de test, pas de boucle, et on pourrait expliciter le comportement pour les 3 états. Je ne sais pas s'il y a plus idiomatique par contre.

Reply:
Alors le truc plus idiomatique, je crois que ça serait les [shared_examples](https://relishapp.com/rspec/rspec-core/docs/example-groups/shared-examples).

Ici, un truc du genre :

```ruby
shared_examples 'inserting the --lien document justificatif-- tag' do
  it { is_expected.to include("Télécharger le document justificatif") }
end

context 'in closed mail' do
  let(:mail) { create(:closed_mail, procedure: procedure) }
  it_behaves_like 'inserting the --lien document justificatif-- tag'
end

context 'in refused mail' do
  let(:mail) { create(:refused_mail, procedure: procedure) }
  it_behaves_like 'inserting the --lien document justificatif-- tag'
end

#…
```

Après, pour revenir sur _« une des autres caractéristiques du --lien attestation-- c'est qu'il n'est dispo que pour le passage en accepté »_, je vois la logique :) Mais en même temps c'est toujours délicat de tester de la logique qui n'est pas là, ou plus là. Sinon, on peut écrire des tests à l'infini ("Est-ce que le comportement marche toujours si la démarche est archivée ? Ou en test ? Ou que l'Usager est aussi Instructeur ?")

Mais comme tu le sens, moins de test ou un shared_exemple, les deux me semblent possible 👍